File: app/src/main/java/com/example/beto/viagemplanejada/AddPublicacao.java
package com.example.beto.viagemplanejada;

import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.RatingBar;
import android.widget.Spinner;
import android.widget.Toast;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;

import org.apache.commons.lang3.StringUtils;

import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class AddPublicacao extends AppCompatActivity implements AdapterView.OnItemSelectedListener {
    FirebaseDatabase firebaseDatabase;

    private EditText dtViagem ;
    private  EditText cidade;

    RatingBar ratingBar;
    Spinner spinner;
    String id;
    List<Pais> paises;
    String paisSelected;
    Publicacao publicacao = new Publicacao();
    List<String> paisesTraduzidos = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_publicacao);
        firebaseDatabase = FirebaseDatabase.getInstance();
        Call<List<Pais>> call;
        id = "";
        dtViagem = findViewById(R.id.editDtViagem);
        cidade = findViewById(R.id.editCidade);
        dtViagem.addTextChangedListener(MaskEditUtil.mask(dtViagem, MaskEditUtil.FORMAT_DATE));
        ratingBar =findViewById(R.id.ratingBar);


        spinner = (Spinner) findViewById(R.id.spinner);


        Bundle extras = getIntent().getExtras();

        if(extras != null){
             id = (String) extras.get("id");
        }

        if(StringUtils.isNotBlank(id)) {
            DatabaseReference databaseRef = firebaseDatabase.getReference().child("Publicacao").child(id);
            databaseRef.addValueEventListener(new ValueEventListener() {
                @Override
                public void onDataChange(DataSnapshot dataSnapshot) {
                    if (dataSnapshot.exists()) {
                        publicacao = dataSnapshot.getValue(Publicacao.class);

                        cidade.setText(publicacao.getCidade());
                        String dtConvert;
                        DateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy");
                        dtConvert = dateFormat.format(publicacao.getDtViagem());
                        dtViagem.setText(dtConvert);
                        ratingBar.setRating(publicacao.getRating());

                    }
                }

                @Override
                public void onCancelled(@NonNull DatabaseError databaseError) {

                }
            });

        }

        try {

            call = new RetrofitConfig().getPaisService().buscarPais();
            call.enqueue(new Callback<List<Pais>>() {
                @Override
                public void onResponse(Call<List<Pais>> call, Response<List<Pais>> response) {

                   paises =  response.body();




                    for(Pais paisTraduzido : paises){

                        paisesTraduzidos.add(paisTraduzido.getTranslations().getBr());

                    }

                    ArrayAdapter<String> adapter = new ArrayAdapter(getApplicationContext(),
                            android.R.layout.simple_spinner_item,
                            paisesTraduzidos);
                    adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
                    spinner.setOnItemSelectedListener(AddPublicacao.this);
                    int spinnerPostion = adapter.getPosition(publicacao.getPais());

                    spinner.setAdapter(adapter);
                    spinner.setSelection(spinnerPostion);


                }

                @Override
                public void onFailure(Call<List<Pais>> call, Throwable t) {
                    Log.e("PaisService   ", "Erro ao buscar o pais:" + t.getMessage());
                }
            });
        }catch (Exception e){
            Log.e("RetrofitConfig", e.getMessage());
        }







    }

    public void savePublicacao(View view) throws ParseException {

        ratingBar.getRating();
        Boolean bTemErro = false;

        if(StringUtils.isBlank(dtViagem.getText().toString())) {
            dtViagem.setError("Data da viagem é um campo obrigatório");
            bTemErro = true;
        }
        if(StringUtils.isBlank(cidade.getText().toString())) {
            cidade.setError("Cidade da viagem é um campo obrigatório");
            bTemErro = true;
        }

        if(!bTemErro) {
            Date dtConvert = new Date();
            DateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy");
            dtConvert = dateFormat.parse(dtViagem.getText().toString());

            DatabaseReference databaseRef = firebaseDatabase.getReference();

            Publicacao publicacao = new Publicacao(paisSelected,
                    cidade.getText().toString(),
                    dtConvert,
                    new Date(), ratingBar.getRating());

            if(StringUtils.isBlank(id)){
                databaseRef.child("Publicacao").push().setValue(publicacao);

            }else{
                databaseRef.child("Publicacao").child(id).setValue(publicacao);
            }
            finish();
        }
    }

    @Override
    public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
        paisSelected = (String) adapterView.getItemAtPosition(i);


    }

    @Override
    public void onNothingSelected(AdapterView<?> adapterView) {


    }

}

File: app/src/main/java/com/example/beto/viagemplanejada/AddPublicacao.kt
package com.example.beto.viagemplanejada


import android.app.Application
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import com.example.beto.viagemplanejada.model.Pais
import com.example.beto.viagemplanejada.model.Publicacao
import com.example.beto.viagemplanejada.services.RetrofitConfig

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener



import java.text.ParseException
import java.text.SimpleDateFormat


import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class AddPublicacao(): AppCompatActivity(), AdapterView.OnItemSelectedListener {
    internal lateinit var firebaseDatabase: FirebaseDatabase

    private var dtViagem: EditText? = null
    private var cidade: EditText? = null

    internal lateinit var ratingBar: RatingBar
    internal lateinit var spinner: Spinner
    internal lateinit var id: String
    internal var paises: List<Pais>? = null
    internal lateinit var paisSelected: String
    internal var publicacao: Publicacao? = Publicacao()
    internal var paisesTraduzidos: MutableList<String> = ArrayList()
    private lateinit var addPublicacaoViewModel: AddPublicacaoViewModel

    override fun onCreate(savedInstanceState: Bundle?) {

        super.onCreate(savedInstanceState)
        addPublicacaoViewModel = ViewModelProviders.of(this)
                .get(AddPublicacaoViewModel::class.java)

        val extras = intent.extras
        setContentView(R.layout.activity_add_publicacao)
        firebaseDatabase = FirebaseDatabase.getInstance()
        val call: Call<List<Pais>>
        id = ""
        dtViagem = findViewById(R.id.editDtViagem)
        cidade = findViewById(R.id.editCidade)
        dtViagem!!.addTextChangedListener(MaskEditUtil.mask(dtViagem!!, MaskEditUtil.FORMAT_DATE))



        spinner = findViewById<View>(R.id.spinner) as Spinner

        addPublicacaoViewModel.errorMessage.observe(this, Observer {
            Toast.makeText(this, it, Toast.LENGTH_SHORT).show()
        })



        if (extras != null) {
            id = extras.get("id") as String
        }

        if (id.isNotBlank()) {
            val databaseRef = firebaseDatabase.reference.child("Publicacao").child(id)
            databaseRef.addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if (dataSnapshot.exists()) {
                        publicacao = dataSnapshot.getValue<Publicacao>(Publicacao::class.java!!)

                        cidade!!.setText(publicacao!!.cidade)
                        val dtConvert: String
                        val dateFormat = SimpleDateFormat("dd/MM/yyyy")
                        dtConvert = dateFormat.format(publicacao!!.dtViagem)
                        dtViagem!!.setText(dtConvert)
                        ratingBar.rating = publicacao!!.rating

                    }
                }

                override fun onCancelled(databaseError: DatabaseError) {

                }
            })

        }

        try {

            call = RetrofitConfig.paisService.buscarPais()
            call.enqueue(object : Callback<List<Pais>> {
                override fun onResponse(call: Call<List<Pais>>, response: Response<List<Pais>>) {

                    paises = response.body()




                    for (paisTraduzido in paises!!) {

                        paisesTraduzidos.add(paisTraduzido.translations!!.br!!)

                    }

                    val adapter = ArrayAdapter(getApplication(),
                            android.R.layout.simple_spinner_item,
                            paisesTraduzidos)
                    adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
                    spinner.onItemSelectedListener = this@AddPublicacao
                    val spinnerPostion = adapter.getPosition(publicacao!!.pais)

                    spinner.adapter = adapter
                    spinner.setSelection(spinnerPostion)


                }

                override fun onFailure(call: Call<List<Pais>>, t: Throwable) {
                    Log.e("PaisService   ", "Erro ao buscar o pais:" + t.message)
                }
            })
        } catch (e: Exception) {
            Log.e("RetrofitConfig", e.message)
        }


    }

    @Throws(ParseException::class)
    fun savePublicacao(view: View) {

        ratingBar.rating
        var bTemErro: Boolean = false

        if (dtViagem!!.text.toString().isBlank()) {
            dtViagem!!.error = "Data da viagem é um campo obrigatório"
            bTemErro = true
        }
        if (cidade!!.text.toString().isBlank()) {
            cidade!!.error = "Cidade da viagem é um campo obrigatório"
            bTemErro = true
        }

        if ((!bTemErro)) {
            var dtConvert = ""
            val dateFormat = SimpleDateFormat("dd/MM/yyyy")
            dtConvert = dateFormat.parse(dtViagem!!.text.toString()).toString()

            val databaseRef = firebaseDatabase.reference

            val publicacao = Publicacao(paisSelected,
                    cidade!!.text.toString(),
                    dtConvert,
                    "", ratingBar.rating)

            if (id.isBlank()) {
                databaseRef.child("Publicacao").push().setValue(publicacao)

            } else {
                databaseRef.child("Publicacao").child(id).setValue(publicacao)
            }
            finish()
        }
    }

    override fun onItemSelected(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
        paisSelected = adapterView.getItemAtPosition(i) as String


    }

    override fun onNothingSelected(adapterView: AdapterView<*>) {


    }

}
